packages/salesforcedx-vscode-core/src/commands/forceSourceDeploySourcePath.ts

```diff
@@ -51,79 +49,85 @@ export class LibraryDeploySourcePathExecutor extends DeployExecutor<
   }
 }
 
-export class LibraryPathsGatherer implements ParametersGatherer<string[]> {
-  private uris: vscode.Uri[];
-  public constructor(uris: vscode.Uri[]) {
-    this.uris = uris;
-  }
-  public async gather(): Promise<ContinueResponse<string[]>> {
-    const sourcePaths = this.uris.map(uri => uri.fsPath);
-    return {
-      type: 'CONTINUE',
-      data: sourcePaths
-    };
-  }
-}
-
-export async function forceSourceDeploySourcePath(sourceUri: vscode.Uri) {
+export const forceSourceDeploySourcePaths = async (
+  sourceUri: vscode.Uri | vscode.Uri[] |undefined,
+  uris: vscode.Uri[] | undefined
+) => {
   if (!sourceUri) {
-    const editor = vscode.window.activeTextEditor;
-    if (editor && editor.document.languageId !== 'forcesourcemanifest') {
-      sourceUri = editor.document.uri;
-    } else {
-      const errorMessage = nls.localize(
-        'force_source_deploy_select_file_or_directory'
-      );
-      telemetryService.sendException(
-        'force_source_deploy_with_sourcepath',
-        errorMessage
-      );
-      notificationService.showErrorMessage(errorMessage);
-      channelService.appendLine(errorMessage);
-      channelService.showChannelOutput();
+    // When the source is deployed via the command palette, both sourceUri and uris are
+    // each undefined, and sourceUri needs to be obtained from the active text editor.
+    sourceUri = getUriFromActiveEditor();
+    if (!sourceUri) {
       return;
     }
   }
 
-  const messages: ConflictDetectionMessages = {
-    warningMessageKey: 'conflict_detect_conflicts_during_deploy',
-    commandHint: input => {
-      return new SfdxCommandBuilder()
-        .withArg('force:source:deploy')
-        .withFlag('--sourcepath', input)
-        .build()
-        .toString();
+  // When a single file is selected and "Deploy Source from Org" is executed,
+  // sourceUri is passed, and the uris array contains a single element, the same
+  // path as sourceUri.
+  //
+  // When multiple files are selected and "Deploy Source from Org" is executed,
+  // sourceUri is passed, and is the path to the first selected file, and the uris
+  // array contains an array of all paths that were selected.
+  //
+  // When editing a file and "Deploy This Source from Org" is executed,
+  // sourceUri is passed, but uris is undefined.
+  if (!uris || uris.length < 1) {
+    if (Array.isArray(sourceUri)) {
+      // When "Push-or-deploy-on-save" is enabled, the first parameter
+      // passed in (sourceUri) is actually an array and not a single URI.
+      uris = sourceUri;
+    } else {
+      uris = [];
+      uris.push(sourceUri);
     }
-  };
-
-  const commandlet = new SfdxCommandlet(
-    new SfdxWorkspaceChecker(),
-    new FilePathGatherer(sourceUri),
-    new LibraryDeploySourcePathExecutor(),
-    new CompositePostconditionChecker(
```

Comment:
@jag-sfdc The implementations of the old "single" function vs. the old "multiple" function vary.  The single version calls
```
new CompositePostconditionChecker(
  new SourcePathChecker(),
    new TimestampConflictChecker(false, messages)
)
```

...yet the multiple version calls
```
new TimestampConflictChecker(false, messages)
```

We should discuss this on Monday.

Reply:
I think this is a mistake that we skipped this. We should add new SourcePathChecker() to make sure the code is in packageDirectories directory defined in sfdx-project.json